src/data_structures/fmindex.rs

```diff
@@ -328,55 +376,78 @@ impl FMDIndex {
         self.backward_ext(&interval.swapped(), comp_a)
             .swapped()
     }
-
-    /// Provide a reference to the underlying FMIndex.
-    pub fn fmindex(&self) -> &FMIndex {
-        &self.fmindex
-    }
 }
 
 
 #[cfg(test)]
 mod tests {
     use super::*;
     use alphabets::dna;
-    use data_structures::suffix_array::suffix_array;
-    use data_structures::bwt::bwt;
+    use data_structures::suffix_array::{suffix_array, SuffixArray};
+    use data_structures::bwt::{bwt, less, Occ};
+    use std::rc::Rc;
+
+    fn sa_interval_to_vec(sa: &SuffixArray, interval: &Interval) -> Vec<usize> {
+        (interval.lower..interval.upper).map(|pos| sa.get(pos).unwrap()).collect()
+    }
 
     #[test]
     fn test_smems() {
         let orig_text = b"GCCTTAACAT";
         let revcomp_text = dna::revcomp(orig_text);
         let text_builder: Vec<&[u8]> = vec![orig_text, b"$", &revcomp_text[..], b"$"];
         let text = text_builder.concat();
-        let pos = suffix_array(&text);
-        println!("pos {:?}", pos);
-        println!("text {:?}", text);
-        let fmdindex = FMDIndex::new(bwt(&text, &pos), 3);
+
+        let alphabet = dna::n_alphabet();
+        let sa = Rc::new(suffix_array(&text));
+        let bwt = Rc::new(bwt(&text, sa.as_ref()));
+        let less = Rc::new(less(&bwt, &alphabet));
+        let occ = Rc::new(Occ::new(&bwt, 3, &alphabet));
+
+        let fmindex = FMIndex::new(bwt, less, occ);
```

Comment:
If we implement Deref for Occ, then above line should also work without having Rc wrappers, right (because Vec already implements Deref)? This means that there is no change in usability to the current implementation, but if people need it, they can also use Rc or other wrappers.